File: SistemasProgramacionRFCVB/SistemasProgramacionRFCVB/Formularios/Persona.vb
Public Class Persona
    Public nombre As String
    Public apellidopaterno As String
    Public apellidomaterno As String
    Public fechanacimiento As DateTime

    'Declaración de propiedades
    Public Property NombreP() As String
        Get  'obtiene los valores de los atributos
            Return nombre
        End Get
        Set(value As String)
            nombre = value
        End Set
    End Property

    Public Property ApellidoPaternoP() As String
        Get  'obtiene los valores de los atributos
            Return apellidopaterno
        End Get
        Set(value As String)
            apellidopaterno = value
        End Set
    End Property

    Public Property ApellidoMaternoP() As String
        Get  'obtiene los valores de los atributos
            Return apellidomaterno
        End Get
        Set(value As String)
            apellidomaterno = value
        End Set
    End Property

    Public Property FechaNacimientoP() As DateTime
        Get  'obtiene los valores de los atributos
            Return fechanacimiento
        End Get
        Set(value As DateTime)
            fechanacimiento = value
        End Set
    End Property

    'Método constructor
    Public Sub Persona(ByVal nombrePersona As String, ByVal apPaternoPersona As String, ByVal apMaternoPersona As String, ByVal fechaNacimientoPersona As DateTime)
        nombre = nombrePersona
        apellidopaterno = apPaternoPersona
        apellidomaterno = apMaternoPersona
        fechanacimiento = fechaNacimientoPersona
    End Sub

    Public ReadOnly Property ObtenerRFC() As String
        Get
            'Dim digito1, digito3, digito4 As String
            'Dim digito2 As Char

            'digito1 = Mid(apellidopaterno, 1, 1)
            ''apellidopaterno.Substring(0, 1)
            ''digito2 = apellidopaterno.Chars(1)
            'digito3 = apellidomaterno.Substring(0, 1)
            'digito4 = nombre.Substring(0, 1)

            ''Si el digito es una Ñ, reemplaza por una X
            'If digito1 = "Ñ" Then
            '    digito1 = "X"
            'End If

            'If digito2 = "Ñ" Then
            '    digito2 = "X"
            'End If

            ''Dígito 2: Toma la primera vocal del primer apellido, si no pone una X
            'Dim paterno() As Char = apellidopaterno.ToCharArray()
            'Dim vocales() As Char = {"A"c, "E"c, "I"c, "O"c, "U"c}

            'Dim i, j As Integer
            'j = 1

            'For i = 0 To vocales.Length Step 1
            '    If vocales(i) = paterno(j) Then
            '        digito2 = paterno(j)
            '    End If
            '    j += 1
            '    While (j < paterno.Length And digito2 <> "A" And digito2 <> "E" And digito2 <> "i" And digito2 <> "u" And digito2 <> "U")
            '        If vocales(i) = paterno(j) Then
            '            digito2 = paterno(j)
            '        End If
            '        j += 1
            '    End While
            '    j = 0
            'Next i

            Dim rfc As String
            'rfc = nombre + apellidopaterno + ApellidoMaternoP + NombreP + FechaNacimientoP.ToString("yyMMdd")
            rfc = NombreP
            Return rfc
        End Get
    End Property
End Class

File: SistemasProgramacionRFC/SistemasProgramacionRFC/Clases/Persona.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace SistemasProgramacionRFC.Clases
{
    class Persona
    {
        public static string nombre;
        public static string apellidopaterno;
        public static string apellidomaterno;
        public static DateTime fechanacimiento;
        public static string rfc = "";

        public string Nombre
        {
            get { return nombre; }
            set { nombre = value; }
        }

        public string ApellidoPaterno
        {
            get { return apellidopaterno; }
            set { apellidopaterno = value; }
        }

        public string ApellidoMaterno
        {
            get { return apellidomaterno; }
            set { apellidomaterno = value; }
        }

        public DateTime FechaNacimiento
        {
            get { return fechanacimiento; }
            set { fechanacimiento = value; }
        }

        //Método constructor
        public Persona(string nombrePersona,
            string apPaternoPersona,
            string apMaternoPersona,
            DateTime fechaNacimientoPersona)
        {
            this.Nombre = nombrePersona;
            this.ApellidoPaterno = apPaternoPersona;
            this.ApellidoMaterno = apMaternoPersona;
            this.FechaNacimiento = fechaNacimientoPersona;
        }
        
        //Propiedad de solo lectura que regresa el rfc
        public static string RFC
        {
            get
            {
                string digito1 = apellidopaterno.Substring(0, 1);
                char digito2 = apellidopaterno[1];
                string digito3 = apellidomaterno.Substring(0, 1);
                string digito4 = nombre.Substring(0, 1);

                //Si el digito es una Ñ, reemplaza por una X
                if (digito1 == "Ñ")
                {
                    digito1 = "X";
                }
                if (digito2 == 'Ñ')
                {
                    digito2 = 'X';
                }
                //Dígito 2: Toma la primera vocal del primer apellido, si no pone una X
                char[] paterno = apellidopaterno.ToCharArray();
                char[] vocales = { 'A', 'E', 'I', 'O', 'U' };

                int j = 1;
                for (int i = 0; i < vocales.Length; i++)
                {
                    if (vocales[i] == digito2)
                    {
                        digito2 = paterno[j];
                    }
                    j++;
                    while (j < paterno.Length && digito2 != 'A' && digito2 != 'E' && digito2 != 'I' && digito2 != 'O' && digito2 != 'U')
                    {
                        if (vocales[i] == paterno[j])
                        {
                            digito2 = paterno[j];
                        }
                        j++;
                    }
                    j = 0;
                }
                return rfc = digito1 + digito2 + digito3 + digito4 + fechanacimiento.ToString("yyMMdd");                
            }
        }
    }
}
